File: src/agents/alerter/src/farmos/logs.js
'use strict';

// Phase 40 D-03 / D-03c: B7 log creation. Native types only (C5):
// seeding | activity | input | observation | harvest. Any other log_type
// throws UnsupportedLogTypeError BEFORE any farmOS call (commit-router catches).
//
// Phase 48 Plan 01: 'seeding_session' is a COMPOSITE log_type recognized by the
// commit-router (so the guard does not bounce it) but NOT a native farmOS log
// type. createLog rejects it -- the seeding_session handler (Plan 02) writes
// one asset + N child seeding logs by calling createLog(client, 'seeding', ...).
// NATIVE_LOG_TYPES is the createLog allow-list; LOG_TYPES is the router guard.
//
// Phase 51 UPSERT-02: upsertLog(client, type, opts) -- lookup-by-stable-key
// then PATCH-merge-or-POST-create. Only 'seeding' migrates this phase (B5
// invariant: one seeding log per child asset makes (type='seeding',
// asset.id == assetIds[0]) an unambiguous stable key). All other log types
// map to null in LOG_STABLE_KEYS and fall through to createLog POST behavior.

const { STABLE_NOTES_SEPARATOR } = require('./merge');

const NATIVE_LOG_TYPES = ['seeding', 'activity', 'input', 'observation', 'harvest'];
const LOG_TYPES = [...NATIVE_LOG_TYPES, 'seeding_session'];

class UnsupportedLogTypeError extends Error {
  constructor(logType) {
    super('unsupported_log_type:' + logType);
    this.name = 'UnsupportedLogTypeError';
    this.logType = logType;
  }
}

class LogIdentityCollision extends Error {
  constructor(logType, assetId, matchedIds) {
    super('log_identity_collision:' + logType + ':' + assetId);
    this.name = 'LogIdentityCollision';
    this.logType = logType;
    this.assetId = assetId;
    this.matchedIds = matchedIds;
  }
}

// Phase 51 UPSERT-02: per-type stable-key resolvers. Only 'seeding' migrates
// to upsert in this phase (B5 invariant: one seeding log per child asset).
// Other types map to null -> POST-only path preserved.
const LOG_STABLE_KEYS = {
  seeding: ({ assetIds } = {}) => (assetIds && assetIds[0]
    ? { path: '/api/log/seeding?filter[asset.id][value]=' + encodeURIComponent(assetIds[0]) }
    : null),
  activity: null,
  input: null,
  observation: null,
  harvest: null,
};

function _buildNoteValue(notes, draftId) {
  return (notes ? notes + '\n' : '') + 'mushy:draft:' + draftId;
}

function _buildLogBody(logType, opts) {
  const { name, timestamp, assetIds = [], fileIds = [], notes = '', draftId } = opts;
  const payload = {
    data: {
      type: 'log--' + logType,
      attributes: {
        name,
        timestamp: Math.floor(timestamp),
        status: 'done',
        notes: { value: _buildNoteValue(notes, draftId), format: 'plain_text' },
      },
      relationships: {
        asset: { data: assetIds.map((id) => ({ type: 'asset--fungi', id })) },
      },
    },
  };
  if (fileIds && fileIds.length > 0) {
    payload.data.relationships.file = {
      data: fileIds.map((id) => ({ type: 'file--file', id })),
    };
  }
  return payload;
}

async function createLog(client, logType, opts) {
  if (!NATIVE_LOG_TYPES.includes(logType)) {
    throw new UnsupportedLogTypeError(logType);
  }
  const payload = _buildLogBody(logType, opts);
  const r = await client.post('/api/log/' + logType, payload);
  if (!r.ok) {
    return { ok: false, reason: 'http_' + (r.status || 'network'), http_status: r.status };
  }
  const logId = r.body && r.body.data && r.body.data.id;
  return { ok: true, logId, http_status: r.status };
}

// ---- Phase 51 UPSERT-02 helpers ----

function _setUnionRefs(existingArr, incomingArr) {
  const existing = Array.isArray(existingArr) ? existingArr : [];
  const incoming = Array.isArray(incomingArr) ? incomingArr : [];
  const byId = new Map();
  for (const ref of existing) {
    if (ref && ref.id != null && !byId.has(ref.id)) byId.set(ref.id, ref);
  }
  for (const ref of incoming) {
    if (ref && ref.id != null && !byId.has(ref.id)) byId.set(ref.id, ref);
  }
  return Array.from(byId.values());
}

function _mergeNotes(existing, incoming) {
  const ev = (existing && existing.value) || '';
  const iv = (incoming && incoming.value) || '';
  const sep = STABLE_NOTES_SEPARATOR;
  const entries = ev.split(sep).map((s) => s.trim()).filter((s) => s.length > 0);
  for (const e of iv.split(sep).map((s) => s.trim()).filter((s) => s.length > 0)) {
    if (!entries.includes(e)) entries.push(e);
  }
  return { value: entries.join(sep), format: 'plain_text' };
}

function _arraysEqualById(a, b) {
  const aa = Array.isArray(a) ? a : [];
  const bb = Array.isArray(b) ? b : [];
  if (aa.length !== bb.length) return false;
  const aIds = aa.map((r) => r && r.id).sort();
  const bIds = bb.map((r) => r && r.id).sort();
  for (let i = 0; i < aIds.length; i++) if (aIds[i] !== bIds[i]) return false;
  return true;
}

function _sortMatches(matches) {
  return matches.slice().sort((a, b) => {
    const ca = (a.attributes && a.attributes.created) || '';
    const cb = (b.attributes && b.attributes.created) || '';
    if (ca !== cb) return ca < cb ? -1 : 1;
    if (a.id < b.id) return -1;
    if (a.id > b.id) return 1;
    return 0;
  });
}

async function _emitAudit(auditLogger, event, payload) {
  if (auditLogger && typeof auditLogger.logCommit === 'function') {
    try { await auditLogger.logCommit(event, payload); } catch (_) { /* non-fatal */ }
  }
}

async function upsertLog(client, type, opts) {
  // Reject non-native types up front (mirror createLog).
  if (!NATIVE_LOG_TYPES.includes(type)) {
    throw new UnsupportedLogTypeError(type);
  }

  const keyFn = LOG_STABLE_KEYS[type];

  // Pass-through: types with null stable key go straight to POST.
  if (keyFn === null) {
    const r = await createLog(client, type, opts);
    if (r.ok) return Object.assign({}, r, { outcome: 'created', conflicts: [], etag_source: null });
    return r;
  }

  const key = keyFn(opts);
  if (key === null) {
    return { ok: false, reason: 'missing_stable_key', http_status: null };
  }

  // Lookup by stable key.
  const lookup = await client.get(key.path);
  if (!lookup.ok) {
    return { ok: false, reason: 'http_' + (lookup.status || 'network'), http_status: lookup.status };
  }
  const rawMatches = (lookup.body && lookup.body.data) || [];

  // Miss: POST via createLog.
  if (rawMatches.length === 0) {
    const r = await createLog(client, type, opts);
    if (r.ok) return Object.assign({}, r, { outcome: 'created', conflicts: [], etag_source: null });
    return r;
  }

  // Hit (>=1 match): sort, pick oldest, surface collision warning if >1.
  const sorted = _sortMatches(rawMatches);
  const canonical = sorted[0];
  const warnings = [];
  if (sorted.length > 1) {
    const matchedIds = sorted.map((m) => m.id);
    warnings.push('LogIdentityCollision:' + sorted.length);
    await _emitAudit(opts.auditLogger, 'log_identity_collision', {
      log_type: type,
      asset_id: opts.assetIds && opts.assetIds[0],
      matched_ids: matchedIds,
    });
  }

  // GET full body for merge.
  const fullResp = await client.get('/api/log/' + type + '/' + canonical.id);
  if (!fullResp.ok) {
    return { ok: false, reason: 'http_' + (fullResp.status || 'network'), http_status: fullResp.status, warnings };
  }
  const existing = fullResp.body && fullResp.body.data;
  const preMergeRevisionId = existing && existing.attributes && existing.attributes.drupal_internal__revision_id;

  // Build incoming and merge.
  const incomingPayload = _buildLogBody(type, opts);
  const incoming = incomingPayload.data;

  // Identity check: asset.data must match between existing and incoming (the stable key).
  const existingAssetData = (existing.relationships && existing.relationships.asset && existing.relationships.asset.data) || [];
  const incomingAssetData = incoming.relationships.asset.data || [];
  if (!_arraysEqualById(existingAssetData, incomingAssetData)) {
    return {
      ok: false,
      reason: 'log_identity_mismatch',
      http_status: null,
      logId: canonical.id,
      warnings,
    };
  }

  // Merge file.data: set-union by id.
  const existingFileData = (existing.relationships && existing.relationships.file && existing.relationships.file.data) || [];
  const incomingFileData = (incoming.relationships.file && incoming.relationships.file.data) || [];
  const mergedFiles = _setUnionRefs(existingFileData, incomingFileData);
  const filesChanged = mergedFiles.length !== existingFileData.length;

  // Merge notes (split-dedup-join).
  const mergedNotes = _mergeNotes(
    existing.attributes && existing.attributes.notes,
    incoming.attributes.notes
  );
  const notesChanged = !(existing.attributes && existing.attributes.notes &&
    existing.attributes.notes.value === mergedNotes.value);

  // Scalar conflicts: timestamp / status / name -- equal=noop, differ=conflict.
  const conflicts = [];
  for (const field of ['timestamp', 'status', 'name']) {
    const ev = existing.attributes && existing.attributes[field];
    const iv = incoming.attributes[field];
    if (iv == null) continue;
    if (ev != null && ev !== iv) {
      conflicts.push({ field, existing: ev, incoming: iv, kind: 'scalar_conflict' });
    }
  }

  if (!filesChanged && !notesChanged) {
    return {
      ok: true,
      logId: canonical.id,
      outcome: 'noop',
      conflicts,
      etag_source: 'soft_compare',
      http_status: null,
      warnings,
    };
  }

  // Build PATCH body with merged file + notes (preserve asset identity).
  const patchBody = {
    data: {
      type: 'log--' + type,
      id: canonical.id,
      attributes: {
        notes: mergedNotes,
      },
      relationships: {
        asset: { data: existingAssetData },
        file: { data: mergedFiles },
      },
    },
  };

  // PATCH with If-Match; soft-compare retry once on 412.
  const ifMatch = preMergeRevisionId != null ? String(preMergeRevisionId) : null;
  const patchOpts = ifMatch ? { headers: { 'If-Match': ifMatch } } : undefined;

  let patchResp = await client.patch('/api/log/' + type + '/' + canonical.id, patchBody, patchOpts);
  if (!patchResp.ok && patchResp.status === 412) {
    // Soft-compare retry: re-GET, rebuild merge, PATCH once more.
    const reGet = await client.get('/api/log/' + type + '/' + canonical.id);
    if (!reGet.ok) {
      return { ok: false, reason: 'http_' + (reGet.status || 'network'), http_status: reGet.status, warnings };
    }
    const refreshed = reGet.body && reGet.body.data;
    const refreshedRev = refreshed && refreshed.attributes && refreshed.attributes.drupal_internal__revision_id;
    const refreshedFiles = (refreshed.relationships && refreshed.relationships.file && refreshed.relationships.file.data) || [];
    const refreshedNotes = refreshed.attributes && refreshed.attributes.notes;
    const remergeFiles = _setUnionRefs(refreshedFiles, incomingFileData);
    const remergeNotes = _mergeNotes(refreshedNotes, incoming.attributes.notes);
    const retryBody = {
      data: {
        type: 'log--' + type,
        id: canonical.id,
        attributes: { notes: remergeNotes },
        relationships: {
          asset: { data: existingAssetData },
          file: { data: remergeFiles },
        },
      },
    };
    const retryHeaders = refreshedRev != null ? { headers: { 'If-Match': String(refreshedRev) } } : undefined;
    patchResp = await client.patch('/api/log/' + type + '/' + canonical.id, retryBody, retryHeaders);
  }

  if (!patchResp.ok) {
    return { ok: false, reason: 'http_' + (patchResp.status || 'network'), http_status: patchResp.status, warnings };
  }

  return {
    ok: true,
    logId: canonical.id,
    outcome: 'patched',
    conflicts,
    etag_source: 'soft_compare',
    http_status: patchResp.status,
    warnings,
  };
}

module.exports = {
  LOG_TYPES,
  NATIVE_LOG_TYPES,
  UnsupportedLogTypeError,
  LogIdentityCollision,
  LOG_STABLE_KEYS,
  createLog,
  upsertLog,
};

File: src/farm-agent/farm_agent/farmos/logs.py
"""farm_agent/farmos/logs.py -- Faithful Python port of the Node farmOS log layer.

Source-of-truth: src/agents/alerter/src/farmos/logs.js

Phase 40 D-03 / D-03c: B7 log creation. Native types only (C5):
seeding | activity | input | observation | harvest. Any other log_type
raises UnsupportedLogTypeError BEFORE any farmOS call.

Phase 48 Plan 01: 'seeding_session' is a COMPOSITE log_type recognized by the
commit-router (so the guard does not bounce it) but NOT a native farmOS log
type. create_log rejects it -- the seeding_session handler writes one asset +
N child seeding logs by calling create_log(client, 'seeding', ...).
NATIVE_LOG_TYPES is the create_log allow-list; LOG_TYPES is the router guard.

Phase 51 UPSERT-02: upsert_log(client, type, opts) -- lookup-by-stable-key
then PATCH-merge-or-POST-create. Only 'seeding' migrates this phase (B5
invariant: one seeding log per child asset makes (type='seeding',
asset.id == assetIds[0]) an unambiguous stable key). All other log types
map to None in LOG_STABLE_KEYS and fall through to create_log POST behavior.

No em-dashes in source artifacts.
"""
from __future__ import annotations

import math
from urllib.parse import quote as _url_quote

from farm_agent.farmos.merge import STABLE_NOTES_SEPARATOR

NATIVE_LOG_TYPES = ["seeding", "activity", "input", "observation", "harvest"]
LOG_TYPES = [*NATIVE_LOG_TYPES, "seeding_session"]


class UnsupportedLogTypeError(Exception):
    """Raised when a log_type is not in NATIVE_LOG_TYPES."""

    def __init__(self, log_type: str):
        super().__init__("unsupported_log_type:" + log_type)
        self.name = "UnsupportedLogTypeError"
        self.log_type = log_type


class LogIdentityCollision(Exception):
    """Raised (or surfaced as warning) when >1 log matches the stable key."""

    def __init__(self, log_type: str, asset_id: str, matched_ids: list):
        super().__init__("log_identity_collision:" + log_type + ":" + asset_id)
        self.name = "LogIdentityCollision"
        self.log_type = log_type
        self.asset_id = asset_id
        self.matched_ids = matched_ids


# Phase 51 UPSERT-02: per-type stable-key resolvers. Only 'seeding' migrates
# to upsert in this phase (B5 invariant: one seeding log per child asset).
# Other types map to None -> POST-only path preserved.
def _seeding_stable_key(opts):
    asset_ids = (opts or {}).get("asset_ids") or []
    if not asset_ids:
        return None
    return {"path": "/api/log/seeding?filter[asset.id][value]=" + _url_quote(str(asset_ids[0]))}


LOG_STABLE_KEYS = {
    "seeding": _seeding_stable_key,
    "activity": None,
    "input": None,
    "observation": None,
    "harvest": None,
}


def _build_note_value(notes: str, draft_id: str) -> str:
    """Append mushy:draft:<id> marker. Mirrors _buildNoteValue()."""
    return (notes + "\n" if notes else "") + "mushy:draft:" + draft_id


def _build_log_body(log_type: str, opts: dict) -> dict:
    """Build the JSON:API payload for a log. Mirrors _buildLogBody()."""
    name = opts.get("name")
    timestamp = opts.get("timestamp", 0)
    asset_ids = opts.get("asset_ids") or []
    file_ids = opts.get("file_ids") or []
    notes = opts.get("notes") or ""
    draft_id = opts.get("draft_id", "")

    payload: dict = {
        "data": {
            "type": "log--" + log_type,
            "attributes": {
                "name": name,
                "timestamp": math.floor(timestamp),
                "status": "done",
                "notes": {
                    "value": _build_note_value(notes, draft_id),
                    "format": "plain_text",
                },
            },
            "relationships": {
                "asset": {
                    "data": [{"type": "asset--fungi", "id": aid} for aid in asset_ids]
                },
            },
        }
    }
    if file_ids:
        # MUSHY-131: 'image', NOT 'file'. farmOS answers 422 to a jpg on the
        # file relationship, which would have failed the ENTIRE log creation,
        # not just the photo. This never fired in prod only because ctx was
        # None and file_ids was therefore always empty (commit_watchdog.py).
        # Settled on dev 2026-08-29: file rel -> 422, image rel -> 201.
        payload["data"]["relationships"]["image"] = {
            "data": [{"type": "file--file", "id": fid} for fid in file_ids]
        }
    return payload


async def create_log(client: dict, log_type: str, opts: dict) -> dict:
    """POST a new log to farmOS. Mirrors createLog().

    Returns {"ok", "log_id", "http_status"} on success, or
    {"ok": False, "reason": "http_<status|network>"} on failure.
    Raises UnsupportedLogTypeError for non-native types (before any call).
    """
    if log_type not in NATIVE_LOG_TYPES:
        raise UnsupportedLogTypeError(log_type)

    payload = _build_log_body(log_type, opts)
    r = await client["post"]("/api/log/" + log_type, payload)
    if not r.get("ok"):
        status = r.get("status")
        return {
            "ok": False,
            "reason": "http_" + (str(status) if status is not None else "network"),
            "http_status": status,
        }
    log_id = None
    body = r.get("body")
    if body and body.get("data"):
        log_id = body["data"].get("id")
    return {"ok": True, "log_id": log_id, "http_status": r.get("status")}


# ---- Phase 51 UPSERT-02 helpers ----


def _set_union_refs(existing_arr, incoming_arr):
    """Set-union of JSON:API ref objects by id. Mirrors _setUnionRefs()."""
    existing = existing_arr if isinstance(existing_arr, list) else []
    incoming = incoming_arr if isinstance(incoming_arr, list) else []
    by_id = {}
    for ref in existing:
        if ref and ref.get("id") is not None and ref["id"] not in by_id:
            by_id[ref["id"]] = ref
    for ref in incoming:
        if ref and ref.get("id") is not None and ref["id"] not in by_id:
            by_id[ref["id"]] = ref
    return list(by_id.values())


def _merge_notes(existing_notes, incoming_notes):
    """Split-dedup-join on STABLE_NOTES_SEPARATOR. Mirrors _mergeNotes()."""
    ev = (existing_notes or {}).get("value") or ""
    iv = (incoming_notes or {}).get("value") or ""
    sep = STABLE_NOTES_SEPARATOR
    entries = [s.strip() for s in ev.split(sep) if s.strip()]
    for e in (s.strip() for s in iv.split(sep) if s.strip()):
        if e not in entries:
            entries.append(e)
    return {"value": sep.join(entries), "format": "plain_text"}


def _arrays_equal_by_id(a, b):
    """True iff both arrays contain the same set of ids. Mirrors _arraysEqualById()."""
    aa = a if isinstance(a, list) else []
    bb = b if isinstance(b, list) else []
    if len(aa) != len(bb):
        return False
    a_ids = sorted(r.get("id") for r in aa if r)
    b_ids = sorted(r.get("id") for r in bb if r)
    return a_ids == b_ids


def _sort_matches(matches):
    """Sort log stubs oldest-first by attributes.created, then id. Mirrors _sortMatches()."""
    def _key(m):
        created = (m.get("attributes") or {}).get("created") or ""
        return (created, m.get("id") or "")
    return sorted(matches, key=_key)


async def _emit_audit(audit_logger, event, payload):
    """Non-fatal audit call. Mirrors _emitAudit()."""
    if audit_logger and callable(audit_logger.get("log_commit")):
        try:
            await audit_logger["log_commit"](event, payload)
        except Exception:
            pass


async def upsert_log(client: dict, log_type: str, opts: dict) -> dict:
    """Lookup-by-stable-key then PATCH-merge-or-POST-create. Mirrors upsertLog().

    Returns a result dict with keys:
      ok, log_id, outcome, conflicts, etag_source, http_status, warnings, reason

    Raises UnsupportedLogTypeError for non-native types (before any call).
    """
    if log_type not in NATIVE_LOG_TYPES:
        raise UnsupportedLogTypeError(log_type)

    key_fn = LOG_STABLE_KEYS[log_type]

    # Pass-through: types with None stable key go straight to POST.
    if key_fn is None:
        r = await create_log(client, log_type, opts)
        if r.get("ok"):
            return {**r, "outcome": "created", "conflicts": [], "etag_source": None}
        return r

    key = key_fn(opts)
    if key is None:
        return {"ok": False, "reason": "missing_stable_key", "http_status": None}

    # Lookup by stable key.
    lookup = await client["get"](key["path"])
    if not lookup.get("ok"):
        status = lookup.get("status")
        return {
            "ok": False,
            "reason": "http_" + (str(status) if status is not None else "network"),
            "http_status": status,
        }
    raw_matches = (lookup.get("body") or {}).get("data") or []

    # Miss: POST via create_log.
    if not raw_matches:
        r = await create_log(client, log_type, opts)
        if r.get("ok"):
            return {**r, "outcome": "created", "conflicts": [], "etag_source": None}
        return r

    # Hit (>=1 match): sort, pick oldest, surface collision warning if >1.
    sorted_matches = _sort_matches(raw_matches)
    canonical = sorted_matches[0]
    warnings = []
    if len(sorted_matches) > 1:
        matched_ids = [m.get("id") for m in sorted_matches]
        warnings.append("LogIdentityCollision:" + str(len(sorted_matches)))
        await _emit_audit(
            opts.get("audit_logger"),
            "log_identity_collision",
            {
                "log_type": log_type,
                "asset_id": (opts.get("asset_ids") or [None])[0],
                "matched_ids": matched_ids,
            },
        )

    # GET full body for merge.
    full_resp = await client["get"]("/api/log/" + log_type + "/" + canonical["id"])
    if not full_resp.get("ok"):
        status = full_resp.get("status")
        return {
            "ok": False,
            "reason": "http_" + (str(status) if status is not None else "network"),
            "http_status": status,
            "warnings": warnings,
        }
    existing = (full_resp.get("body") or {}).get("data")
    pre_merge_revision_id = (
        (existing or {}).get("attributes", {}).get("drupal_internal__revision_id")
    )

    # Build incoming and merge.
    incoming_payload = _build_log_body(log_type, opts)
    incoming = incoming_payload["data"]

    # Identity check: asset.data must match (the stable key).
    existing_asset_data = (
        ((existing or {}).get("relationships") or {}).get("asset", {}).get("data") or []
    )
    incoming_asset_data = incoming["relationships"]["asset"]["data"] or []
    if not _arrays_equal_by_id(existing_asset_data, incoming_asset_data):
        return {
            "ok": False,
            "reason": "log_identity_mismatch",
            "http_status": None,
            "log_id": canonical["id"],
            "warnings": warnings,
        }

    # Merge the photo relationship: set-union by id.
    # MUSHY-131: 'image', matching the create path. farmOS 422s a jpg on 'file'.
    existing_file_data = (
        ((existing or {}).get("relationships") or {}).get("image", {}).get("data") or []
    )
    incoming_file_data = (
        (incoming.get("relationships") or {}).get("image", {}).get("data") or []
    )
    merged_files = _set_union_refs(existing_file_data, incoming_file_data)
    files_changed = len(merged_files) != len(existing_file_data)

    # Merge notes (split-dedup-join).
    merged_notes = _merge_notes(
        (existing or {}).get("attributes", {}).get("notes"),
        incoming["attributes"]["notes"],
    )
    existing_notes_value = (
        ((existing or {}).get("attributes") or {}).get("notes", {}) or {}
    ).get("value")
    notes_changed = existing_notes_value != merged_notes["value"]

    # Scalar conflicts: timestamp / status / name -- equal=noop, differ=conflict.
    conflicts = []
    for field in ("timestamp", "status", "name"):
        ev = ((existing or {}).get("attributes") or {}).get(field)
        iv = incoming["attributes"].get(field)
        if iv is None:
            continue
        if ev is not None and ev != iv:
            conflicts.append({"field": field, "existing": ev, "incoming": iv, "kind": "scalar_conflict"})

    if not files_changed and not notes_changed:
        return {
            "ok": True,
            "log_id": canonical["id"],
            "outcome": "noop",
            "conflicts": conflicts,
            "etag_source": "soft_compare",
            "http_status": None,
            "warnings": warnings,
        }

    # Build PATCH body with merged file + notes (preserve asset identity).
    patch_body = {
        "data": {
            "type": "log--" + log_type,
            "id": canonical["id"],
            "attributes": {
                "notes": merged_notes,
            },
            "relationships": {
                "asset": {"data": existing_asset_data},
                "image": {"data": merged_files},
            },
        }
    }

    # PATCH with If-Match; soft-compare retry once on 412.
    if_match = str(pre_merge_revision_id) if pre_merge_revision_id is not None else None
    patch_opts = {"headers": {"If-Match": if_match}} if if_match else None

    patch_resp = await client["patch"](
        "/api/log/" + log_type + "/" + canonical["id"], patch_body, patch_opts
    )

    if not patch_resp.get("ok") and patch_resp.get("status") == 412:
        # Soft-compare retry: re-GET, rebuild merge, PATCH once more.
        re_get = await client["get"]("/api/log/" + log_type + "/" + canonical["id"])
        if not re_get.get("ok"):
            status = re_get.get("status")
            return {
                "ok": False,
                "reason": "http_" + (str(status) if status is not None else "network"),
                "http_status": status,
                "warnings": warnings,
            }
        refreshed = (re_get.get("body") or {}).get("data")
        refreshed_rev = ((refreshed or {}).get("attributes") or {}).get("drupal_internal__revision_id")
        refreshed_files = (
            ((refreshed or {}).get("relationships") or {}).get("image", {}).get("data") or []
        )
        refreshed_notes = ((refreshed or {}).get("attributes") or {}).get("notes")
        remerge_files = _set_union_refs(refreshed_files, incoming_file_data)
        remerge_notes = _merge_notes(refreshed_notes, incoming["attributes"]["notes"])
        retry_body = {
            "data": {
                "type": "log--" + log_type,
                "id": canonical["id"],
                "attributes": {"notes": remerge_notes},
                "relationships": {
                    "asset": {"data": existing_asset_data},
                    "image": {"data": remerge_files},
                },
            }
        }
        retry_headers = (
            {"headers": {"If-Match": str(refreshed_rev)}} if refreshed_rev is not None else None
        )
        patch_resp = await client["patch"](
            "/api/log/" + log_type + "/" + canonical["id"], retry_body, retry_headers
        )

    if not patch_resp.get("ok"):
        status = patch_resp.get("status")
        return {
            "ok": False,
            "reason": "http_" + (str(status) if status is not None else "network"),
            "http_status": status,
            "warnings": warnings,
        }

    return {
        "ok": True,
        "log_id": canonical["id"],
        "outcome": "patched",
        "conflicts": conflicts,
        "etag_source": "soft_compare",
        "http_status": patch_resp.get("status"),
        "warnings": warnings,
    }
